app/src/main/java/org/thoughtcrime/securesms/mediasend/CameraContactsRepository.java

```diff
@@ -0,0 +1,136 @@
+import android.text.TextUtils;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.WorkerThread;
+
+import org.signal.core.util.concurrent.SignalExecutors;
+import org.signal.core.util.logging.Log;
+import org.thoughtcrime.securesms.contacts.ContactRepository;
+import org.thoughtcrime.securesms.database.DatabaseFactory;
+import org.thoughtcrime.securesms.database.GroupDatabase;
+import org.thoughtcrime.securesms.database.RecipientDatabase;
+import org.thoughtcrime.securesms.database.ThreadDatabase;
+import org.thoughtcrime.securesms.database.model.ThreadRecord;
+import org.thoughtcrime.securesms.recipients.Recipient;
+import org.thoughtcrime.securesms.recipients.RecipientId;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+/**
+ * Handles retrieving the data to be shown in {@link CameraContactSelectionFragment}.
+ */
+class CameraContactsRepository {
+
+  private static final String TAG = Log.tag(CameraContactsRepository.class);
+
+  private static final int RECENT_MAX = 25;
+
+  private final Context           context;
+  private final ThreadDatabase    threadDatabase;
+  private final GroupDatabase     groupDatabase;
+  private final RecipientDatabase recipientDatabase;
+  private final ContactRepository contactRepository;
+  private final Executor          serialExecutor;
+  private final ExecutorService   parallelExecutor;
+
+  CameraContactsRepository(@NonNull Context context) {
+    this.context           = context.getApplicationContext();
+    this.threadDatabase    = DatabaseFactory.getThreadDatabase(context);
+    this.groupDatabase     = DatabaseFactory.getGroupDatabase(context);
+    this.recipientDatabase = DatabaseFactory.getRecipientDatabase(context);
+    this.contactRepository = new ContactRepository(context);
+    this.serialExecutor    = SignalExecutors.SERIAL;
+    this.parallelExecutor  = SignalExecutors.BOUNDED;
+  }
+
+  void getCameraContacts(@NonNull Callback<CameraContacts> callback) {
+    getCameraContacts("", callback);
+  }
+
+  void getCameraContacts(@NonNull String query, @NonNull Callback<CameraContacts> callback) {
+    serialExecutor.execute(() -> {
+      Future<List<Recipient>> recents  = parallelExecutor.submit(() -> getRecents(query));
+      Future<List<Recipient>> contacts = parallelExecutor.submit(() -> getContacts(query));
+      Future<List<Recipient>> groups   = parallelExecutor.submit(() -> getGroups(query));
+
+      try {
+        long           startTime = System.currentTimeMillis();
+        CameraContacts result    = new CameraContacts(recents.get(), contacts.get(), groups.get());
+
+        Log.d(TAG, "Total time: " + (System.currentTimeMillis() - startTime) + " ms");
+
+        callback.onComplete(result);
+      } catch (InterruptedException | ExecutionException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)